### ISSUE TITLE
E035: Failed to create the following bookmark on the server: [object Object]

### ISSUE THREAD

**Issue by tom-ch1** [floccusaddon/floccus#1799]
### Which version of floccus are you using?

5.4.0

### How many bookmarks do you have, roughly?

800

### Are you using other means to sync bookmarks in parallel to floccus?

No

### Sync method

Nextcloud Bookmarks

### Which browser are you using? In case you are using the phone App, specify the Android or iOS version and device please.

Chromium Version 131.0.6778.139 (Official Build) built on Debian GNU/Linux 12 (bookworm) (64-bit)

### Which version of Nextcloud Bookmarks are you using? (if relevant)

v14.2.4

### Which version of Nextcloud? (if relevant)

Nextcloud Hub 7 (28.0.9)

### Describe the Bug

I have 3 sync tasks: a) for the bookmark bar, b) for other bookmarks, and c) for Tabs.
a) and b) are working fine but c) stops with an error: 
**Syncing failed with E035: Failed to create the following bookmark on the server: [object Object]**

### Expected Behavior

Syncing Tabs should succeed

### To Reproduce

It is unclear what causes the issue. From previous experiences with floccus - nextcloud, it was always due to long page names or special characters. As floccus does not tell me which tab fails (is that a bug in the logger?) and only outputs **[object Object]** it is impossible for me to check this out.

### Debug log provided

- [ ] I have provided a debug log file

**Comment by tom-ch1** [comment]
Donation done :-)

**Comment by marcelklehr** [comment]
Hello :wave: 
Can you check the logs for the full error message?

**Comment by marcelklehr** [comment]
Can you still send the last logs (possibly in redacted form)? I might be able to figure it out from there.

**Comment by tom-ch1** [comment]
Hi Marcel, I found the redacted log and redacted it even more with:
* replaced my server with https://my.redacted.nextcloud.domain
* ```sed 's/\(url=https\?%3A%2F%2F\).*$/\1-redacted/' floccus-5.4.0-2024-12-15-redacted.log > floccus-5.4.0-2024-12-15-really-redacted.log``` because of some leaking bookmark urls.

You'll find the log on my google drive: https://drive.google.com/file/d/1sb0lWWzuahcx26T4lNAnfSYpZwFMcXIZ/view?usp=sharing

**Comment by tom-ch1** [comment]
I don't understand: There are dozens of bookmarks created not just 6. What do I search for in the log to find the 6 you are talking about?

**Comment by marcelklehr** [comment]
Sorry, you're right, it's not just 6, but right before the error and after `Server: executing CREATEs` I count 13 log entries with `(nextcloud-folders)CREATE`, which signals the execution of a create action. That should hopefully be easier to pin down than all bookarks.

> How can I install a version with this commit? https://github.com/floccusaddon/floccus/commit/9b2f6b9a8006a17df261f201906ba2a5bfd2d4ec

Yes, that would be easier. You or someone else needs to build it.\* I expect to work on floccus during the holidays and will probably publish this patch along with the next version soon, but I cannot promise anything.

\* Build steps can be found here: https://github.com/floccusaddon/floccus?tab=readme-ov-file#development

**Comment by tom-ch1** [comment]
Thanks Marcel, I built floccus and installed the extension. I found the bookmarks which were causing the errors. They have long names, but I can't definitely determine the problem. I then closed all the failed tabs an re-ran a Sync. Now there are other errors (maybe related?): 

```
2024-12-22T23:32:27.790Z Syncing failed with Cannot read properties of undefined (reading 'bookmark')
Cannot read properties of undefined (reading 'bookmark')
... (9 other rows with the same message)```

**Comment by marcelklehr** [comment]
Can you send logs again of this new error?

**Comment by marcelklehr** [comment]
I've published a new alpha version for you to try, which should fix the "Failed to create bookmark" error: https://github.com/floccusaddon/floccus/releases/tag/v5.4.2-alpha.1

**Comment by tom-ch1** [comment]
@marcelklehr thanks for the alpha release. It provides better localisation of the erroneous Tab. I'll try to provide a chronology of relevant log entries for the following Tab which triggers this issue:
* https://github.com/BIPES/BIPES

Logfile extract:
```
2024-12-29T12:51:34.045Z { localTreeRoot: 
     + #tabs[] parentId: undefined, hash: undefined
...
       + #369564945[Window 17] parentId: tabs, hash: undefined
...
         - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945
...
  serverTreeRoot: 
     + #9556[Open_Tabs_Xindi] parentId: undefined, hash: fd3841321b7aaab559720c1cab8df722cdb64015519b9c6a1c019e7e8d766c26
...
       + #9774[Window 17] parentId: 9556, hash: undefined
... --> note that in this section, the tab in question does not occur because it could not be created in the last run
  cacheTreeRoot: 
     + #tabs[undefined] parentId: undefined, hash: undefined
    }
2024-12-29T12:51:34.050Z Calculating diff for      + #9556[Open_Tabs_Xindi] parentId: undefined, hash: fd3841321b7aaab559720c1cab8df722cdb64015519b9c6a1c019e7e8d766c26
...
      + #9774[Window 17] parentId: 9556, hash: undefined
... --> no mention of the offending tab
      + #369564945[Window 17] parentId: tabs, hash: undefined
...
        - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945
...
2024-12-29T12:51:34.070Z Calculating diff for      + #9774[Window 17] parentId: 9556, hash: undefined
... --> no mention of offending tab, missing newline after the last entry, hence it continues directly with:
     + #369564945[Window 17] parentId: tabs, hash: undefined
...
      - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945
... --> in this section no mentions of offending tab, apparently only creates log-entry for elements which are present locally and remote...
2024-12-29T12:51:34.071Z Scanner: Finding moves
2024-12-29T12:51:34.072Z Scanner: Generate reorders
2024-12-29T12:51:34.073Z Calculating diff for      + #tabs[] parentId: undefined, hash: undefined
...
      + #369564945[Window 17] parentId: tabs, hash: undefined
...
        - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945
...
      + #9774[Window 17] parentId: 9556, hash: undefined
...
2024-12-29T12:51:34.085Z Calculating diff for      + #369564945[Window 17] parentId: tabs, hash: undefined
...
      - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945
... --> missing newline
     + #9774[Window 17] parentId: 9556, hash: undefined
... --> in this section no mentions of offending tab, apparently only creates log-entry for elements which are present locally and remote...
2024-12-29T12:51:34.088Z Scanner: Finding moves
2024-12-29T12:51:34.088Z Scanner: Generate reorders
2024-12-29T12:51:34.103Z { localScanResult: 
   { CREATE: 
      Diff
...
      Action: CREATE
      Payload: #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945 Index: 1
      ,
...
     REORDER: 
      Diff
...
      Action: REORDER
      Payload: #369564945[Window 17] parentId: tabs Order: [
... --> mentions the order of local Tab IDs, the offending tab is mentioned
      Action: REORDER
      Payload: #tabs[] parentId: undefined Order: [
... --> mentions the order of local Window IDs 
      ] },
  serverScanResult: 
   { CREATE: 
      Diff
... --> what? offending tab not mentioned?
     REMOVE: 
      Diff
...
      Action: REMOVE
      Payload: #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945 Index: 37
... --> it is unclear (and incorrect) why the diff is resolved by "removing" it locally instead of adding it remote
      Action: REORDER
      Payload: #9774[Window 17] parentId: 9556 Order: [
... --> does not mention offending tab
      Action: REORDER
      Payload: #9556[Open_Tabs_Xindi] parentId: undefined Order: [
... --> mentions the order of remote Window IDs 
      ] } }
2024-12-29T12:51:34.107Z { revertPlan: 
   { CREATE: 
      Diff
...
      Action: CREATE
      Payload: #undefined[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 9774 Index: 1
...
      Action: REORDER
      Payload: #9774[Window 17] parentId: 9556 Order: [
... --> no mention of the offending tab
2024-12-29T12:51:34.108Z Executing Server revert plan
2024-12-29T12:51:34.108Z Executing revert plan for Server
2024-12-29T12:51:34.108Z Server: executing CREATEs
...
2024-12-29T12:51:34.116Z Executing action  { type: 'CREATE',
  payload:   - #undefined[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 9774,
  index: 1,
  oldItem:   - #369564984[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 369564945 }
2024-12-29T12:51:34.116Z (nextcloud-folders)CREATE     - #undefined[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 9774
2024-12-29T12:51:34.117Z QUEUING GET https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark?url=https%3A%2F%2Fgithub.com%2FBIPES%2FBIPES
2024-12-29T12:51:34.117Z FETCHING GET https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark?url=https%3A%2F%2Fgithub.com%2FBIPES%2FBIPES
... --> many unrelated entries (it would be nice to have a debugging option which forces serial blocking execution)
2024-12-29T12:51:34.559Z Receiving response for GET https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark?url=https%3A%2F%2Fgithub.com%2FBIPES%2FBIPES
2024-12-29T12:51:34.561Z QUEUING POST https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark
2024-12-29T12:51:34.561Z FETCHING POST https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark
2024-12-29T12:51:34.571Z Receiving response for POST https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark
2024-12-29T12:51:34.572Z Executed 2 actions from 39 actions
... --> it is unclear which elements are queued / fetched /posted / received / executed
2024-12-29T12:51:36.659Z POST https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark: Server responded with 400: {"status":"error","data":["Could not add bookmark"]}
2024-12-29T12:51:36.749Z Receiving response for POST https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/bookmark
2024-12-29T12:51:36.754Z Syncing failed with E035: Failed to create the following bookmark on the server: - #undefined[BIPES/BIPES: BIPES: Block based Integrated Platform for Embedded Systems allows text and block based programming for several types of embedded systems and Internet of Things modules using MicroPython, CircuitPython, Python or Snek. You can connect, program, debug and monitor several types of boards using network, USB or Bluetooth. No software install needed!](https://github.com/BIPES/BIPES) parentId: 9774
E035: Failed to create the following bookmark on the server: - #undefined[jczic/MicroWebSrv2: The last Micro Web Server for IoTs (MicroPython) or large servers (CPython), that supports WebSockets, routes, template engine and with really optimized architecture (mem allocations, async I/Os). Ready for ESP32, STM32 on Pyboard, Pycom's chipsets (WiPy, LoPy, ...). Robust, efficient and documented!](https://github.com/jczic/MicroWebSrv2) parentId: 9774
... --> strange, the second E035 error has a different log entry. The timestamp and "Syncing failed with " is missing
2024-12-29T12:51:36.756Z QUEUING DELETE https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/lock
2024-12-29T12:51:36.757Z FETCHING DELETE https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/lock
2024-12-29T12:51:36.829Z Receiving response for DELETE https://my.nextcloud.url/index.php/apps/bookmarks/public/rest/v2/lock
... --> I don't know if the DELETEs are related to this issue.
```
For what it's worth: There seem to be higher risks to run into this issue on github project pages. May it be related to the lengthy description? It would be interesting to see if this failing url also fails for other nextcloud bookmark installations.

**Comment by tom-ch1** [comment]
@marcelklehr : I highly suspect that the underlying cause for this issue is a database field length restriction in the nextcloud app. Try the following:
1. start with a fresh floccus profile that syncs a certain folder to your nextcloud server
2. open https://github.com/BIPES/BIPES and add it as a bookmark (inside the folder to be synced
3. make a floccus upload once -> for me, this results in an error
4. edit the bookmark and shorten the title
5. make a floccus upload once -> for me, this results runs successful

So I'd like to suggest 2 new features:
1. after a failed create, retry with a truncated title
2. continue after failed operations

For bookmarks, I can shorten the title as a workaround. For Tab sync, this unfortunately won't work

**Comment by marcelklehr** [comment]
Mh. It could be that you still have a very old database schema in Nextcloud Bookmarks, apparently I need to fix this, you're not the first with this issue. It can happen when you had installed bookmarks on your nextcloud prior to database migrations being introduced. What is the field length of oc_bookmarks.title for you?

**Comment by tom-ch1** [comment]
it's ```character varying(255)```, should probably be ```text``` I suppose.
Is there a documentation of the database schema in nc bookmarks? I could check if there are other discrepancies...

**Comment by marcelklehr** [comment]
Yes, it should have length 4092

**Comment by tom-ch1** [comment]
isn't "text" better than "character varying(4092)"? currently, my url and description are text and user_id is "character varying(64)"

**Comment by tom-ch1** [comment]
I went through all "lib/Migration/Version*.php" files and compared it to my db schema (that was a pain...)
Turns out that only bookmarks.title was wrong. I have a faint memory that I manually changed that column due to an earlier bug. This might have caused the migration to fail somehow :-(

Anyway: Thanks, @marcelklehr , I could resolve this issue by altering the db column. And as doctrine convert 4096 to infinity, I will not ever run into this issue again (which might not be true for other users who will hit it when somebody adds a fancy long description to a github project...)
